render/src/shape_utils.rs

```diff
@@ -1283,3 +1171,149 @@ impl<'a> LineScales<'a> {
+        let commands = ShapeConverter::from_shape(&shape).into_commands();
+        let expected = vec![DrawPath::Fill {
+            style: &FILL_STYLES[0],
+            commands: vec![
+                DrawCommand::MoveTo(swf::Point::from_pixels(100.0, 100.0)),
+                DrawCommand::LineTo(swf::Point::from_pixels(200.0, 100.0)),
+                DrawCommand::LineTo(swf::Point::from_pixels(200.0, 200.0)),
+                DrawCommand::LineTo(swf::Point::from_pixels(100.0, 200.0)),
+                DrawCommand::LineTo(swf::Point::from_pixels(100.0, 100.0)),
+            ],
+            winding_rule: FillRule::EvenOdd,
+        }];
+        assert_eq!(commands, expected);
+    }
+
+    /// A solid square with one edge flipped (fillstyle0 instead of fillstyle1).
+    #[test]
+    fn flipped_edges() {
+        let shape = build_shape(vec![
+            ShapeRecord::StyleChange(Box::new(swf::StyleChangeData {
+                move_to: Some(swf::Point::from_pixels(100.0, 100.0)),
+                fill_style_0: None,
+                fill_style_1: Some(1),
+                line_style: None,
+                new_styles: None,
+            })),
+            ShapeRecord::StraightEdge {
+                delta: PointDelta::from_pixels(100.0, 0.0),
+            },
+            ShapeRecord::StraightEdge {
+                delta: PointDelta::from_pixels(0.0, 100.0),
+            },
+            ShapeRecord::StraightEdge {
+                delta: PointDelta::from_pixels(-100.0, 0.0),
+            },
+            ShapeRecord::StyleChange(Box::new(swf::StyleChangeData {
+                move_to: Some(swf::Point::from_pixels(100.0, 100.0)),
+                fill_style_0: Some(1),
+                fill_style_1: Some(0),
+                line_style: None,
+                new_styles: None,
+            })),
+            ShapeRecord::StraightEdge {
+                delta: PointDelta::from_pixels(0.0, 100.0),
+            },
+        ]);
+        let commands = ShapeConverter::from_shape(&shape).into_commands();
+        let expected = vec![DrawPath::Fill {
+            style: &FILL_STYLES[0],
+            commands: vec![
+                DrawCommand::MoveTo(swf::Point::from_pixels(100.0, 100.0)),
+                DrawCommand::LineTo(swf::Point::from_pixels(200.0, 100.0)),
+                DrawCommand::LineTo(swf::Point::from_pixels(200.0, 200.0)),
+                DrawCommand::LineTo(swf::Point::from_pixels(100.0, 200.0)),
+                DrawCommand::LineTo(swf::Point::from_pixels(100.0, 100.0)),
+            ],
+            winding_rule: FillRule::EvenOdd,
+        }];
+        assert_eq!(commands, expected);
+    }
+
+    use swf::Twips;
+
+    #[test]
+    fn test_winding_number_line() {
+        fn test(
+            test_point: swf::Point<Twips>,
+            begin: swf::Point<Twips>,
+            end: swf::Point<Twips>,
+            expected: i32,
+        ) {
+            let result = crate::shape_utils::winding_number_line(test_point, begin, end);
```

Comment:
nit: Actually I'm almost sure that a fully qualified name is unnecessary here. It can be referenced as simply as `winding_number_line()`.

Reply:
You are right, when I first wrote it, I tried without any qualification, and that did not work for some reason, but when I try now, it works.